### ISSUE TITLE
Vault dynamic secrets expire when service account lease is renewed

### ISSUE THREAD

**Issue by nimjor** [external-secrets/external-secrets#2198]
**Describe the solution you'd like**
Using kubernetes auth method with a role and service account with access to Hashicorp Vault with dynamic secrets generator, credentials generated are tied to the service account's token, and are revoked by Vault when that token expires.  If this happens before the next `ExternalSecret.spec.refreshInterval`, then the `Kind=Secret` target will hold expired secret values until the refreshInterval occurs.

Basically, unless the ExternalSecret sourced from a VaultDynamicSecret is refreshed at the same time as the service account renews its lease against Vault, there will be some period where the secret data is stale/invalid.

Is there a way already and I've just missed it, or can we add a way, for ExternalSecrets referencing dynamic secrets to be refreshed based on some other trigger rather than a static interval?  In other words, it seems important to have the refresh triggered based on some other metadata (service account's Vault lease renewal time or something) to avoid this out of sync state.

**What is the added value?**
Dynamic secrets will be fresh and usable at any time; their k8s Secret resources will never hold stale values.

**Give us examples of the outcome**
`time=0s`: k8s SA `foo` granted Vault token X with lease TTL 3600s
`time=10s`: VaultDynamicSecret + ExternalSecret `bar` created using kubernetes auth for SA `foo`, secret `refreshInterval: 3600s`
`time=3600s`: SA `foo` renews Vault lease, next expiration at time=7200s, existing secrets tied to old lease are revoked
`time=3605s`: Pod `baz` tries to launch using ES `bar` for image pull :skull_and_crossbones: ErrImagePull
`time=3610s`: ES `bar` refreshes dynamic secret values, next expiration at time=7210s...

**Observations (Constraints, Context, etc):**
What we are trying to accomplish, in this case, is dynamic (essentially secret-less) auth against Artifactory for imagePullSecrets.  We are using [JFrog's Vault Artifactory Secrets Plugin](https://github.com/jfrog/artifactory-secrets-plugin) which enables these dynamic secrets.

Here are the relevant resource manifests:
```yaml
apiVersion: generators.external-secrets.io/v1alpha1
kind: VaultDynamicSecret
metadata:
  name: vault-dynamic
  namespace: demo
spec:
  path: "/artifactory/token/demo"
  method: "GET"
  provider:
    server: "https://vault-dev.tools.demo.com"
    auth:
      kubernetes:
        mountPath: "k8s-platform"
        role: "demo-tools-access-read"
        serviceAccountRef:
          name: "demo"
---
apiVersion: external-secrets.io/v1beta1
kind: ExternalSecret
metadata:
  name: image-pull-dynamic
  namespace: demo
spec:
  refreshInterval: "3600s"
  target:
    name: image-pull-dynamic
    template:
      type: kubernetes.io/dockerconfigjson
      engineVersion: v2
      data:
        .dockerconfigjson: |
          {
            "auths": {
              "containers.artifactory-dev.tools.demo.com": {
                "auth": "{{ list .username ":" .access_token | join "" | b64enc }}",
                "email": "demo@example.com",
                "password": "{{ .access_token }}",
                "username": "{{ .username }}"
              }
            }
          }
  dataFrom:
  - sourceRef:
      generatorRef:
        apiVersion: generators.external-secrets.io/v1alpha1
        kind: VaultDynamicSecret
        name: vault-dynamic
```

**Comment by moolen** [comment]
I think the solution should be to refresh the secret before it expires. Hence, the `refreshInterval` should be < lease TTL (better <50%).

I think this is mostly a documentation issue. The lease is not renewed, a new token is generated and valid for the TTL configured in the associated vault role.

Would that be a problem to just increase the `refreshInterval`?

**Comment by nimjor** [comment]
That doesn’t solve the out of sync problem. The main issue is that there is a hierarchy relationship between the created dynamic secrets and the user/role/token that was used to create them. When the SA’s token expires and a new one is requested, any secrets that were generated with the former token also expire. So unless there is a way to inform the ExternalSecret that the token/its secrets have expired, there will always be a possibility for things to be out of sync. Using very short refreshIntervals is a workaround, but not a complete solution—it still leaves a chance that a pod or will need the secret at a moment when it is not valid.

**Comment by moolen** [comment]
Let me rephrase it:

* A new vault token is generated with every reconciliation for every ExternalSecret resource.
* This token is valid for the TTL configured in the associated vault role.
* ESO uses that token to create dynamic secrets from vault

> When the SA’s token expires and a new one is requested, any secrets that were generated with the former token also expire

Does the vault token depend on the Kubernetes Service account Token expiry? I don't think so. IIRC it has a static TTL configured per-role (thats how i read the TTL setting [here in the docs](https://developer.hashicorp.com/vault/api-docs/auth/kubernetes#parameters-1)).

> any secrets that were generated with the former token also expire

That's right, they do expire after the token TTL. However if the `refreshInterval < token TTL` then a new token is created before the old token + secret expires. And a new dynamic secret is issued and stored in a `Kind=Secret`. 


> So unless there is a way to inform the ExternalSecret that the token/its secrets have expired, there will always be a possibility for things to be out of sync

It's just a matter of tuning the `spec.refreshInterval` and Vault Token TTL so this doesn't happen. There is no possibility for it to get out of sync if this is tuned correctly.

This relates to https://github.com/external-secrets/external-secrets/issues/2184#issuecomment-1498183680. Currently, refreshInterval should be less then 50% of the Vault TTL, otherwise you may run into the edge case described in this issue.
